bayes_opt1/util.py

```diff
@@ -0,0 +1,547 @@
+        m = min(bounds[ax,]) # len(bounds[ax,]) = 2 always \forall p
+        M = max(bounds[ax,]) # len(bounds[ax,]) = 2 always \forall p
+        for i in range(R):
+            seq[i,ax] = m + seq[i,ax]*(M-m)
+    return seq
+
+
+# Implementation algorithm 2 (Frazier)
+
+def _kg2(x, optimizer, gp, n_grid = 100, J = 300):
+
+    
+    def x_obs_to_array(d, p):
+        ''' 
+        Transforms a dictionary d into an array out (used STRICTLY for x_obs)
+         - d: input dictionary
+         - p: number of coordinates (e.g. if we are working in R^6, then p = 6)
+        '''
+        out = [np.zeros(p)] * len(d)
+        for i in range(len(d)):
+            # d[i] is a vector with only one element --> d[i][0] is a dictionary
+            v = d[i][0].items() 
+            out[i] = [item[1] for item in v]
+        out = np.array(out)
+        return out
+    
+    def posterior(gp, x_obs, y_obs, grid):
+        '''
+        fits gaussian process on points (x_obs[i], y_obs[i]) and evaluates its mean on each point
+        in grid. Notice that "fit" method implemented in sklearn library is very expensive.
+        '''
+        gp.fit(x_obs, y_obs)
+        mu = gp.predict(grid, return_std=False)
+        return mu
+    
+    def find_mean_max(x_obs,y_obs,gp,grid,mean=None,std=None,i=None):
+        '''
+        If i is None, then it means we are looking for what the paper called \mu^{*}_{n}, otherwise we want to
+        compute \mu^{*}_{n+1}.
+        Remark: i=None <=> mean=None <=> std=None
+        '''
+        if isinstance(i, type(None)): # checks if i is None
+            x_newobs = x_obs
+            y_newobs = y_obs
+        else:
+            y=np.random.normal(loc=mean, scale=std)
+            x_newobs=np.concatenate((x_obs, i.reshape(1,-1)), axis = 0)
+            y_newobs=np.append(y_obs,y) 
+        mu=posterior(gp, x_newobs, y_newobs, grid)
+        return (max(mu),np.argmax(mu))
+    
+    p = optimizer._space.bounds.shape[0] # size of state space (e.g. if we are working on R^3, n = 3)
+    init = [np.zeros(n_grid)] * p # list of p vectors, each of lenght n_grid   
+    x_obs_temp = np.array([[res["params"]] for res in optimizer.res]) 
+    y_obs = np.array([res["target"] for res in optimizer.res])
+    # x_obs needs to be np.array of size (n,p)
+    # y_obs needs to be np.array of size (n,)
+    # see https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html
+    x_obs = x_obs_to_array(x_obs_temp, p)
+    
+
+    for i in range(p): 
+        init[i] = np.linspace(optimizer._space.bounds[i,0], optimizer._space.bounds[i,1], n_grid) # works \forall p
+
+    grid = np.meshgrid(*init) # list with p matrices n_grid x n_grid
+    for g in range(len(grid)):
+        grid[g] = grid[g].reshape(-1,1)
+    grid = np.stack(grid, axis = -1)
+    grid=grid[:,0,:] # array of shape (n_grid*n_grid, p): each row is composed of one element from init[0] and one element from init[1]
```

Comment:
Metterei queste righe in una funzione a parte che poi chiamate qui, per tenere la funzione principale più ordinata, e per poterla riutilizzare in `_kg4()` iinvece che scrivere le stesse righe